File: V7/src/main/java/uk/co/q3c/v7/base/view/component/Breadcrumb.java
/*
 * Copyright (C) 2013 David Sowerby
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */
package uk.co.q3c.v7.base.view.component;

import java.util.ArrayList;
import java.util.List;

import uk.co.q3c.v7.base.navigate.SitemapURIConverter;
import uk.co.q3c.v7.base.navigate.V7Navigator;
import uk.co.q3c.v7.base.view.V7ViewChangeEvent;
import uk.co.q3c.v7.base.view.V7ViewChangeListener;
import uk.co.q3c.v7.i18n.I18NListener;
import uk.co.q3c.v7.i18n.I18NTranslator;

import com.vaadin.ui.HorizontalLayout;

public class Breadcrumb extends HorizontalLayout implements I18NListener, V7ViewChangeListener {

	private final List<BreadcrumbStep> steps = new ArrayList<>();
	private final V7Navigator navigator;
	private final SitemapURIConverter converter;

	protected Breadcrumb(V7Navigator navigator, SitemapURIConverter converter) {
		this.navigator = navigator;
		navigator.addViewChangeListener(this);
		this.converter = converter;
		moveToNavigationState();
	}

	private void moveToNavigationState() {
		converter.nodeChainForUri(navigator.getNavigationState());
	}

	@Override
	public void localeChange(I18NTranslator translator) {
		moveToNavigationState();
	}

	@Override
	public boolean beforeViewChange(V7ViewChangeEvent event) {
		// return false;
		throw new RuntimeException("not yet implemented");
	}

	@Override
	public void afterViewChange(V7ViewChangeEvent event) {
		moveToNavigationState();
	}

	@Override
	public void detach() {
		navigator.removeViewChangeListener(this);
		super.detach();

	}

}

File: src/main/kotlin/uk/q3c/krail/core/view/component/Breadcrumb.kt
/*
 *
 *  * Copyright (c) 2016. David Sowerby
 *  *
 *  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 *  * the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *  *
 *  * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 *  * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 *  * specific language governing permissions and limitations under the License.
 *
 */
package uk.q3c.krail.core.view.component

import com.google.inject.Inject
import com.google.inject.Provider
import com.vaadin.ui.Component
import org.slf4j.LoggerFactory
import uk.q3c.krail.core.guice.uiscope.UIScoped
import uk.q3c.krail.core.navigate.Navigator
import uk.q3c.krail.core.navigate.sitemap.UserSitemap
import uk.q3c.util.guice.SerializationSupport


interface Breadcrumb : Component

@UIScoped
class DefaultBreadcrumb @Inject constructor(
        navigatorProvider: Provider<Navigator>,
        userSitemapProvider: Provider<UserSitemap>,
        serializationSupport: SerializationSupport)

    : NavigationButtonPanel(navigatorProvider, userSitemapProvider, serializationSupport), Breadcrumb {

    override fun build() {
        log.debug("building breadcrumb")
        val nodeChain = sitemapProvider.get().nodeChainFor(navigatorProvider.get().currentNode)
        organiseButtons(nodeChain)
    }

    companion object {
        private val log = LoggerFactory.getLogger(DefaultBreadcrumb::class.java)
    }
}
